packages/web/src/common/store/upload/sagas.js

```diff
@@ -898,262 +902,92 @@ function* uploadCollection(tracks, userId, collectionMetadata, isAlbum) {
-              urlTitle: formatUrlName(track.metadata.title),
-              handle
-            }
-          })
-          if (maybeTrackRes !== undefined && maybeTrackRes !== null) {
-            return maybeTrackRes
+  const audiusLibs = yield call([
+    audiusBackendInstance,
+    audiusBackendInstance.getAudiusLibs
+  ])
+  const tracksWithMetadata = yield all(
+    tracks.map((track) =>
+      call(function* () {
+        // Technically _generateTrackId() is "private". Abusing JS here to get it
+        // before the upload succeeds, and instead pass it through the upload process.
+        const id = yield call([
+          audiusLibs.Track,
+          audiusLibs.Track._generateTrackId
+        ])
+        return {
+          track,
+          metadata: {
+            ...track.metadata,
+            id
           }
-          // lil bit of delay for retries
-          yield delay(1500)
-          retries -= 1
-        }
-
-        throw new Error(`Exhausted retries querying for track ${trackId}`)
-      },
-      function* onSuccess(confirmedTrack) {
-        yield call(responseChan.put, { confirmedTrack })
-      },
-      function* ({ timeout, message }) {
-        yield put(uploadActions.uploadTrackFailed())
-
-        if (timeout) {
-          yield put(uploadActions.singleTrackTimeoutError())
-        } else {
-          yield put(
-            uploadActions.singleTrackUploadError(
-              message,
-              phaseContainer.phase,
-              track.file.size
-            )
-          )
         }
-        yield cancel(dispatcher)
-        yield call(responseChan.put, { error: message })
-      },
-      () => {},
-      UPLOAD_TIMEOUT_MILLIS
+      })
     )
   )
 
-  const { confirmedTrack, error } = yield take(responseChan)
-  const isRejected = error === 'Request failed with status code 403'
+  // Get stem metadatas
+  const stems = yield select(getStems)
+  const updatedStems = updateAndFlattenStems(
+    stems,
+    tracksWithMetadata.map((t) => t.metadata.id)
+  )
+  const allUploads = tracksWithMetadata.concat(updatedStems)
 
-  yield reportResultEvents({
-    numSuccess: error ? 0 : 1,
-    numFailure: error && !isRejected ? 1 : 0,
-    numRejected: isRejected ? 1 : 0,
-    uploadType: 'single_track',
-    errors: error ? [error] : []
+  // Upload tracks and stems parallel together
+  const { trackIds } = yield call(handleUploads, {
+    tracks: allUploads,
+    isCollection: false
   })
 
-  if (error) {
-    if (isRejected) {
-      yield put(
-        make(Name.TRACK_UPLOAD_COMPLETE_UPLOAD, {
-          count: 1,
-          kind: 'tracks'
-        })
+  // Record stem upload events
+  for (let i = tracksWithMetadata.length; i < trackIds.length; i += 1) {
+    const trackId = trackIds[i]
+    const parentTrackId = allUploads[i]?.metadata?.stem_of?.parent_track_id
+    const category = allUploads[i]?.metadata?.stem_of?.category
+    if (!trackIds.includes(parentTrackId)) {
+      console.debug(
```

Comment:
should this be console error so we get details in sentry?

Reply:
maybe.. though maybe the error should be from the place that failed the upload? This is if the stem uploaded successfully and was deleted because the parent was gone

Reply:
kk yeah nvm